radosgw_agent/worker.py

```diff
@@ -99,15 +101,65 @@ def _meta_entry_from_json(entry):
         )
 
 BucketIndexEntry = namedtuple('BucketIndexEntry',
-                              ['object', 'marker', 'timestamp'])
+                              [
+                                  'object',
+                                  'marker',
+                                  'timestamp',
+                                  'op',
+                                  'versioned',
+                                  'ver',
+                                  'name',
+                                  # compatibility with boto objects:
+                                  'VersionedEpoch',
+                                  'version_id',
+                              ])
+
+BucketVer = namedtuple('BucketVer',
+        [
+            'epoch',
+            'pool',
+        ])
+
 
 def _bi_entry_from_json(entry):
+    ver = entry.get('ver', {})
+    entry_ver = BucketVer(
+        ver.get('epoch'),
+        ver.get('pool')
+    )
+
+    # compatibility with boto objects:
+    VersionedEpoch = ver.get('epoch')
+    version_id = entry.get('instance')
+
     return BucketIndexEntry(
         entry['object'],
         entry['op_id'],
         entry['timestamp'],
+        entry.get('op', ''),
+        entry.get('versioned', False),
+        entry_ver,
+        entry['object'],
+        VersionedEpoch,
+        version_id,
         )
 
+
+def filter_versioned_objects(entry):
+    """
+    On incremental sync operations, the log may indicate that 'olh' entries,
+    which should be ignored. So this filter function will check for the
+    different attributes present in an ``entry`` and return only valid ones.
+
+    This should be backwards compatible with older gateways that return log
+    entries that don't support versioning.
+    """
+    # writes or delete 'op' values should be ignored
+    if entry.op not in ['write', 'delete']:
```

Comment:
from the design doc we can check for the versioned attribute of a log entry to determine whether to treat it as versioned (and skip write and delete entries) or let it through